clang/tools/driver/cc1_main.cpp

```diff
@@ -678,10 +696,30 @@ void CompileJobCache::finishComputedResult(CompilerInstance &Clang,
                                 &Clang.getInvocation().getDiagnosticOpts()),
       /*ShouldOwnClient=*/true);
 
-  if (Error E = CacheBackend->finishComputedResult(*ResultCacheKey)) {
+  // Check if we encounter any source that would generate non-deterministic
+  // outputs.
+  bool SkipCache = false;
+  if (Clang.hasPreprocessor() && Clang.isSourceNonReproducible()) {
+    DiagnosticsEngine &Diags = Clang.getDiagnostics();
+    switch (Clang.getPreprocessorOpts().CachingDiagOption) {
+      case CachingDiagKind::None:
+        break;
+      case CachingDiagKind::Warning:
+        Diags.Report(diag::remark_compile_job_cache_skipped)
+            << ResultCacheKey->toString();
+        SkipCache = true;
```

Comment:
I think we should set `SkipCache` outside the switch. That avoids any possibility of storing non-reproducible value to the cache in case someone modifies the diagnostic option or if we hit the `llvm_unreachable` case in a release build somehow.